grove/tomography/utils.py

```diff
@@ -0,0 +1,627 @@
+    """
+    Generate `n` samples of measuring all outcomes of a Quil `program`
+    assuming the assignment probabilities `assignment_probs` by simulating the
+    wave function on a reference-qvm SyncConnection `cxn`
+
+    :param pyquil.quil.Program program: The program.
+    :param int num_samples: The number of samples
+    :param numpy.ndarray assignment_probs: A matrix of assignment probabilities
+    :param reference_qvm.api.SyncConnection cxn: the QVM connection.
+    :return: The resulting sampled outcomes from assignment_probs applied to cxn.
+    """
+    wf = cxn.wavefunction(program)
+    return sample_outcomes(assignment_probs.dot(abs(wf.amplitudes.ravel())**2), num_samples)
+
+
+def make_histogram(samples, ksup):
+    """
+    For a list of samples [s1, s2, ..., sN] taking on integer values from 0 to ksup-1,
+    make a histogram of each integer's outcome and return it.
+
+    :param samples: The samples.
+    :param ksup: The (exclusive) upper bound
+    :return: A histogram of outcomes.
+    """
+    return np.histogram(samples, np.arange(ksup + 1) - 0.5)[0]
+
+
+def estimate_assignment_probs(bitstring_prep_histograms):
+    """
+    Compute the estimated assignment probability matrix for a sequence of single shot histograms
+    obtained by running the programs generated by `basis_state_preps()`.
+
+        bitstring_prep_histograms[i,j] = #number of measured outcomes j when running program i
+
+    The assignment probability is obtained by transposing and afterwards normalizing the columns.
+
+        p[j, i] = Probability to measure outcome j when preparing the state with program i.
+
+    :param list|numpy.ndarray bitstring_prep_histograms: A nested list or 2d array with shape
+    (d, d), where ``d = 2**nqubits`` is the dimension of the Hilbert space. The first axis varies
+    over the state preparation program index, the second axis corresponds to the measured bitstring.
+    :return: The assignment probability matrix.
+    """
+    p = np.array(bitstring_prep_histograms, dtype=float).T
+    p /= p.sum(axis=0)[np.newaxis, :]
+    return p
+
+
+def plot_pauli_transfer_matrix(ptransfermatrix, ax, labels, title):
+    """
+    Visualize the Pauli Transfer Matrix of a process.
+
+    :param numpy.ndarray ptransfermatrix: The Pauli Transfer Matrix
+    :param ax: The matplotlib axes.
+    :param labels: The labels for the operator basis states.
+    :param title: The title for the plot
+    """
+    im = ax.imshow(ptransfermatrix, interpolation="nearest", cmap=rigetti_3_color_cm, vmin=-1,
+                   vmax=1)
+    dim = len(labels)
+    plt.colorbar(im, ax=ax)
+    ax.set_xticks(range(dim))
+    ax.set_xlabel("Input Pauli Operator", fontsize=20)
+    ax.set_yticks(range(dim))
+    ax.set_ylabel("Output Pauli Operator", fontsize=20)
+    ax.set_title(title, fontsize=25)
+    ax.set_xticklabels(labels, rotation=45)
+    ax.set_yticklabels(labels)
+    ax.grid(False)
```

Comment:
Maybe return the `ax`?

Reply:
Well, given that it's an input parameter, it's already available outside the scope

Reply:
@ntezak I agree, but I'm spooked by functions that don't return and just mutate. Is there anything bad with returning?

Reply:
No, it's fine, I don't care strongly about it